bolt-simple-taproot.md

```diff
@@ -0,0 +1,958 @@
+(cooperative closures, etc) BIP 86 is used as well.
+
+#### Taproot Key Path Spends
+
+A taproot key path spend allows one to spend a taproot output without revealing
+the script root, which may or may not exist. 
+
+The witness of a keypath spend is simply a 64 or 65 byte schnorr signature. A
+signature is 64 bytes if the new `SIGHASH_DEFAULT` alias for `SIGHASH_ALL` is
+being used. Otherwise a single byte for the sighash type is used as normal:
+```
+<key_path_sig>
+```
+
+If an output commits to a script root (or uses BIP 86), a private key will need
+to be tweaked in order to generate a valid signature, and negated if the
+resulting output key has an odd y coordinate (see BIP 341 for details).
+
+#### Taproot Script Path Spends
+
+A script path spend is executed when a tapscript leaf is revealed. A script
+path spends has three components: a control block, the script leaf being
+revealed, and the valid witness.
+
+A control block has the following serialization:
+```
+(output_key_y_parity | leaf_version) || internal_key || inclusion_proof
+```
+
+The first segment uses a spare bit of the `leaf_version` to encode the parity
+of the output key, which is checked during control block verification. The
+`internal_key` is just that. The `inclusion_proof` is the series of sibling
+hashes in the path from the revealed leaf all the way up to the root of the
+tree. In practice if one hashes the revealed leaf, and each 32-byte hash of the
+inclusion proof together, they'll reach the script root if the proof was valid.
+
+If only a single leaf is committed to in a tree, then the `inclusion_proof`
+will be absent.
+
+The final witness to spend a script path output is:
+```
+<witness1> ... <witnessN> <leaf_script> <control_block>
+```
+
+### MuSig2
+
+Musig2 is a multi-signature scheme based on schnorr signatures. It allows N
+parties to construct an aggregated public key, and then generate a single
+multi-signature that can only be generated with all the parties of the
+aggregated key. In practice we use this to allow the funding output of
+channel to look just like any other P2TR output.
+
+A musig2 signing flow has 4 stages:
+
+  1. First each party generates a public nonce, and exchanges it with every
+    other party.
+
+  2. Next public keys are exchanged by each party, and combined into a single
+    aggregated public key.
+
+  3. Next, each party generates a _partial_ signature and exchanges it with
+    every other party.
+
+  4. Finally, once all partial signatures are known, they can be combined into a
+    valid schnorr signture, which can be veirfied using BIP 340.
+
+Steps 1&2 may be performed out of order, or concurrently. In our case only two
+parties exist, so as soon as one party knows both partial signatures, they can
+be combined into a final signature.
+
+#### Key Aggregation
+
+We refer to the of `KeyAgg` algorithm [bip-musig2](https://github.com/jonasnick/bips/blob/musig2/bip-musig2.mediawiki)
```

Comment:
> Might affect some part of this proposal but haven't checked it in full yet.

It'll affect the ultimate test vectors, and interop will require both sides being on the same musig2 version (ofc). `btcd` has implemented up to version 0.4.0 so far, and this is also what's packaged in version 0.15.0 of `lnd` w/ an experimental RPC API. 

I think so far only the Python ref implementation is on version 0.8.1, afact secp-zkp is on 0.3.0 (or so).

Reply:
I think by the time this is ready for prime time (cross impl interop, etc) we'll ideally be at musig 1.0. Then we'll just pin that and we can safely make test vectors to fill in everything.

Reply:
I believe we can update all the `bip-musig2` references to `BIP 327` since we're [now at 1.0.0](https://github.com/bitcoin/bips/blob/b4ea0a93aaa469057cd88695c8d56ef92a6734b3/bip-0327.mediawiki?plain=1#L788).